hermes-client/src/main/java/pl/allegro/tech/hermes/client/HermesClient.java

```diff
@@ -1,42 +1,67 @@
+
 import java.net.URI;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicInteger;
-import java.util.function.Function;
 import java.util.function.Predicate;
 
-import static java.util.concurrent.CompletableFuture.completedFuture;
-import static java.util.stream.IntStream.range;
 import static pl.allegro.tech.hermes.client.HermesMessage.hermesMessage;
 
 public class HermesClient {
 
     private final HermesSender sender;
     private final String uri;
     private final Map<String, String> defaultHeaders;
-    private final int retries;
-    private final Predicate<HermesResponse> retryCondition;
     private final AtomicInteger currentlySending = new AtomicInteger(0);
+    private final RetryPolicy retryPolicy;
+    private final ScheduledExecutorService scheduler;
     private volatile boolean shutdown = false;
 
     HermesClient(HermesSender sender,
                  URI uri,
                  Map<String, String> defaultHeaders,
                  int retries,
-                 Predicate<HermesResponse> retryCondition) {
+                 Predicate<HermesResponse> retryCondition,
+                 long retrySleepInMillis,
+                 long maxRetrySleepInMillis) {
+        this(sender, uri, defaultHeaders, retries, retryCondition, retrySleepInMillis,
+                maxRetrySleepInMillis, Executors.newSingleThreadScheduledExecutor());
+    }
+
+    HermesClient(HermesSender sender,
+                 URI uri,
+                 Map<String, String> defaultHeaders,
+                 int retries,
+                 Predicate<HermesResponse> retryCondition,
+                 long retrySleepInMillis,
+                 long maxRetrySleepInMillis,
+                 ScheduledExecutorService scheduler) {
         this.sender = sender;
         this.uri = createUri(uri);
         this.defaultHeaders = Collections.unmodifiableMap(new HashMap<>(defaultHeaders));
-        this.retries = retries;
-        this.retryCondition = retryCondition;
+        this.retryPolicy = createRetryPolicy(retries, retryCondition, retrySleepInMillis, maxRetrySleepInMillis);
+        this.scheduler = scheduler;
+    }
+
+    private RetryPolicy createRetryPolicy(int retries, Predicate<HermesResponse> retryCondition,
+                                          long retrySleepInMillis, long maxRetrySleepInMillis) {
+        RetryPolicy retryPolicy = new RetryPolicy().withMaxRetries(retries)
+                .retryIf(retryCondition::test);
+        if (retrySleepInMillis > 0) {
+            retryPolicy.withBackoff(retrySleepInMillis, maxRetrySleepInMillis, TimeUnit.MILLISECONDS);
```

Comment:
We don't need to test the internals, just that regardless of time, it does the required number of tries using a latch, as in other tests.